src/Lorentz/Contracts/BaseDAO/Proposal.hs

```diff
@@ -0,0 +1,674 @@
+          dip $ dip swap
+          stackType @[Natural, Address, store, Proposal pm, ProposalKey, [Operation]]
+
+        False -> do
+          dup
+          cRejectedProposalReturnValue
+          stackType @["slash_amount" :! Natural, Proposal pm, store, ProposalKey, [Operation]]
+          dip $ do
+            getField #pProposerFrozenToken
+            dip $ getField #pProposer
+            dip $ dip swap
+          stackType @["slash_amount" :! Natural, Natural, Address, store, Proposal pm, ProposalKey, [Operation]]
+          burnSlashAmount
+
+          -- Calculate unfreeze amount
+          swap; dip (fromNamed #slash_amount)
+          sub; isNat
+          ifSome nop $ do
+            -- If the result is negative, we unfreeze 0 tokens.
+            -- The frozen token balance associated with that proposal will be burn
+            -- via 'burnSlashAmount'
+            push (0 :: Natural)
+            stackType @[Natural, Address, store, Proposal pm, ProposalKey, [Operation]]
+
+unfreezeVoterToken
+  :: forall store pm. (NiceParameter pm, StorageC store pm)
+  => [Proposal pm, store, ProposalKey, [Operation]] :-> [Proposal pm, store, ProposalKey, [Operation]]
+unfreezeVoterToken = do
+  dup; dig @2; swap
+  toField #pVoters
+  iter $ do
+    unpair
+    swap
+    unfreeze
+  swap
+
+handleProposalIsOver
+  :: forall store pm. (NiceParameter pm, StorageC store pm)
+  => Config pm -> '[((Bool, ProposalKey), Proposal pm), store, [Operation]] :-> '[store, [Operation]]
+handleProposalIsOver config@Config{..} = do
+  unpair; unpair
+
+  if Holds then do
+    -- | start unfreezing process
+    swap; dip swap
+    stackType @[Proposal pm, store, ProposalKey, [Operation]]
+
+    getField #pUpvotes
+    dip $ getField #pDownvotes
+    add; toNamed #proposalTotalVote
+
+    duupX @3
+    stToField #sQuorumThreshold; toNamed #sQuorumThreshold
+
+    if #sQuorumThreshold <=. #proposalTotalVote then do
+      getFieldNamed #pUpvotes
+      dip $ getFieldNamed #pDownvotes
+      if #pUpvotes >. #pDownvotes then do
+        unfreezeProposerToken config True
+        unfreezeVoterToken
+        cDecisionLambda
+        unpair
+        stackType @[[Operation], store, ProposalKey, [Operation]]
+        dig @3
+        iter cons
+        swap; dip swap
+      else do
+        -- | Reject proposal if (upvote <= downvote)
+        unfreezeProposerToken config False
+        unfreezeVoterToken
+        drop
+
+    else do
+      -- | Reject proposal regardless of upvotes
+      unfreezeProposerToken config False
+      unfreezeVoterToken
+      drop
+
+    stackType @[store, ProposalKey, [Operation]]
+
+    swap
+    stDelete #sProposals
+
+  else do
+    -- | Add the proposalId back to #sProposalKeyListSortByDate
+    -- since we initially set this to []
+    dip drop
+
+    dip $ stGetField #sProposalKeyListSortByDate
+    cons
+    stSetField #sProposalKeyListSortByDate
+
+-- | Update quroum threshold of all ongoing and new proposals.
```

Comment:
Is this description correct? Quorum threshold is one for the whole contract and is updated by another entrypoint. I think this function can be better described.

Reply:
This is wrong comment for `flush`. I removed this, and add proper `flush` description instead.

I also made some changes to `setQuorumTheshold` comment as well.